java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java

```diff
@@ -183,6 +190,59 @@
 @Beta
 public class WebDriverDecorator<T extends WebDriver> {
 
+  protected static class Definition {
+    private final Class<?> decoratedClass;
+    private final Class<?> originalClass;
+
+    public Definition(Decorated<?> decorated) {
+      this.decoratedClass = decorated.getClass();
+      this.originalClass = decorated.getOriginal().getClass();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == null || getClass() != o.getClass()) return false;
+      Definition definition = (Definition) o;
+      return (decoratedClass == definition.decoratedClass)
+          && (originalClass == definition.originalClass);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(decoratedClass, originalClass);
+    }
+  }
+
+  protected static class ProxyFactory<T> {
+    private final Class<? extends T> clazz;
+    private final Field targetLookup;
+
+    private ProxyFactory(Class<? extends T> clazz, Field targetLookup) {
```

Comment:
I would rather limit the clazz to adhere to an interface containing setter and getter for the decorated target. 
See https://github.com/appium/java-client/blob/42480feca9845bbbb569b13d995b80980bf4ca33/src/main/java/io/appium/java_client/proxy/Helpers.java#L135 for an example

Reply:
@mykola-mokhnach this makes things mutch easier to impement, thanks again!